### PR TITLE
[RFC] Add support for Universal Binary JSON

### PR DESCRIPTION
**Pull request by trivialfis**
This is an RFC for adding binary JSON format (https://ubjson.org/) to XGBoost for
serialization to resolve the performance issues.  In the original proposal of revising the
serialization format, one of the reasons we choose JSON is that there are multiple
specifications of binary equivalence.

## Motivation

The serialization can occur in multiple places, including saving model, pickle,
transferring the model between distributed workers, releasing memory at the end of
training (copy).  We have adopted JSON format in the past, which cleanups the
serialization format, but also introduced performance overhead.  As originally proposed,
we will add a binary implementation when needed.  The binary JSON can close the gap
between the current JSON format and our old binary format, allowing us to phase out support
for the old binary models.

## Universal Binary JSON
Universal Binary JSON (https://ubjson.org/) is a specification for serializing JSON
document into the binary format with a focus on performance and ease of implementation.

So far I have found 2 promising candidates.  First is the BSON specification.  It's widely
adopted and well maintained with big projects like mongoDB utilizing it.  However it's
complicated, requires known document size, and most importantly it has a limit on the
total document size as 16 megabytes. (https://docs.mongodb.com/manual/reference/limits/).
Another one is the ubjson adopted in this RFC, which is efficient for array objects but has
lesser implementations.  For instance non of the existing Python implementations have full
support for the specification.  It has the concept of a typed container (array and object),
which particularly suits the need for XGBoost since XGBoost has many arrays including
tree weights, node indices, etc.  Also, it's a relatively simple specification that we can
implement easily.

## Implementation

The serializer is easily added on top of existing JSON infrastructure since the JSON
implementation in XGBoost has a modularized reader and writer.  Also, I have changed model
serialization to using typed containers wherever appropriate.

## Performance

Benchmark is carried out on a model trained on airline dataset with 1000 boosting rounds
and 12 max\_depth.  The following result is a total of 1000 calls of `Booster.__copy__`.  In
total 1e6 tree copies are carried out in the test.

|      | Master             | PR                 |
|:-----|:-------------------|:-------------------|
| Time | 1279.9649624824524 | 53.655569553375244 |

Close https://github.com/dmlc/xgboost/pull/7145
Close https://github.com/dmlc/xgboost/issues/6697
Close https://github.com/dmlc/xgboost/issues/7158
Close https://github.com/dmlc/xgboost/issues/4060

## TO-DOs
- [x] Remove the experimental tag for JSON model format.
- [x] Investigate JVM support.
- [x] More documents.
- [ ] Add support to the SHAP package.
- [x] Add warnings according to the roadmap.
- [x] Roadmap for deprecating the old binary format.

### CODE REVIEW

**Comment by hcho3** [comment]
I approve the idea of adopting UBJSON to speed up serialization.

> Close #7158

Do you plan to implement a method to serialize Booster object as a string or byte-buffer (without using the disk)?

**Comment by trivialfis** [comment]
We already have that feature, in R that's xgb.save.raw

**Comment by trivialfis** [comment]
Hmm .. the SHAP package is using that raw buffer ...

**Comment by hcho3** [comment]
@trivialfis There isn't an equivalent feature in Python.

**Comment by trivialfis** [comment]
@hcho3 It's a function that calls `Learner.SaveModel`.  This PR changes it to use UBJSON from old binary format, which I might change back due to breaking.

**Comment by trivialfis** [comment]
But the idea is that it's quite easy to change the format for that function.

**Comment by hcho3** [comment]
@trivialfis Can we add an optional argument to `save_raw()`, like `save_raw(dump_format='json')` ? We can set the argument default to `binary` for now, but after a while we can change the default to `json` or `ubjson`.

**Comment by trivialfis** [comment]
Yes, we can.  But we need a new C api function.

**Comment by hcho3** [comment]
Got it. Let's add a new C API function.

**Comment by trivialfis** [comment]
Added a new API function along with support for R and Python.

**Comment by trivialfis** [comment]
@hcho3 I plan to gradually remove the support for the old binary model.  A simple roadmap would be

- Release 1.6: Add ubjson to python/R/jvm, emit a warning about the model format.  The jvm part might be more difficult, we will see.
- Release 2.0: Continue the warning.
- Release 2.1: Continue the warning.
- Release 2.2: Remove support for saving to old binary format.  But keep the loading part with a warning.
- Release 2.3: Remove support for loading old binary format.

I will open an issue for the roadmap if it's feasible.

**Comment by hcho3** [comment]
@trivialfis Can you put up a roadmap issue for phasing out the binary serialization format?

**Comment by RAMitchell** [comment]
This looks very nice. Can we expect dask performance to improve? In the past I remember model serialisation in a multi-process setting to be a significant bottleneck.

**Comment by trivialfis** [comment]
> Can we expect dask performance to improve? In the past I remember model serialisation in a multi-process setting to be a significant bottleneck.

We have optimization on dask to avoid model serialization, specifically, users can pass a future object to xgb.dask.predict.  But for training large models this can still be significant, for instance https://github.com/dmlc/xgboost/issues/5474#issuecomment-976561515 .

File: jvm-packages/xgboost4j/src/native/xgboost4j.cpp [hunk 1]
```diff
@@ -662,20 +662,24 @@ JNIEXPORT jint JNICALL Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGBoosterLoadModel
...
+ * Signature: (JLjava/lang/String;[[B)I
  */
-JNIEXPORT jint JNICALL Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGBoosterGetModelRaw
-  (JNIEnv * jenv, jclass jcls, jlong jhandle, jobjectArray jout) {
+JNIEXPORT jint JNICALL Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGBoosterSaveModelToBuffer
+  (JNIEnv * jenv, jclass jcls, jlong jhandle, jstring jformat, jobjectArray jout) {
   BoosterHandle handle = (BoosterHandle) jhandle;
+  const char *format = jenv->GetStringUTFChars(jformat, 0);
   bst_ulong len = 0;
   const char* result;
```
Comment (trivialfis) [inline jvm-packages/xgboost4j/src/native/xgboost4j.cpp:673]: Initialize it.

File: jvm-packages/xgboost4j/src/native/xgboost4j.cpp [hunk 2]
```diff
@@ -662,20 +662,24 @@ JNIEXPORT jint JNICALL Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGBoosterLoadModel
...
-  int ret = XGBoosterGetModelRaw(handle, &len, &result);
-  JVM_CHECK_CALL(ret);
+  xgboost::Json config {xgboost::Object{}};
+  config["format"] = std::string{format};
+  std::string config_str;
+  xgboost::Json::Dump(config, &config_str);
 
+  int ret = XGBoosterSaveModelToBuffer(handle, config_str.c_str(), &len, &result);
```
Comment (trivialfis) [inline jvm-packages/xgboost4j/src/native/xgboost4j.cpp:679]: Early return.

File: src/c_api/c_api.cc [hunk 3]
```diff
@@ -900,45 +894,110 @@ XGB_DLL int XGBoosterPredictFromCUDAColumnar(
...
   } else {
     std::unique_ptr<dmlc::Stream> fi(dmlc::Stream::Create(fname, "r"));
     static_cast<Learner*>(handle)->LoadModel(fi.get());
   }
   API_END();
 }
 
-XGB_DLL int XGBoosterSaveModel(BoosterHandle handle, const char* c_fname) {
+namespace {
+void WarnOldModel() {
```
Comment (trivialfis) [inline src/c_api/c_api.cc:920]: Typo

File: src/c_api/c_api_utils.h [hunk 4]
```diff
@@ -241,5 +241,25 @@ inline void GenerateFeatureMap(Learner const *learner,
 }
 
 void XGBBuildInfoDevice(Json* p_info);
+
+template <typename JT>
+inline auto RequiredArg(Json in, std::string const &key, StringView func) {
```
Comment (trivialfis) [inline src/c_api/c_api_utils.h:246]: Should accept and return references.

File: doc/tutorials/saving_model.rst [hunk 5]
```diff
@@ -2,16 +2,18 @@
...
-the future until JSON format is no-longer experimental and has satisfying performance.
-This tutorial aims to share some basic insights into the JSON serialisation method used in
-XGBoost.  Without explicitly mentioned, the following sections assume you are using the
-JSON format, which can be enabled by providing the file name with ``.json`` as file
-extension when saving/loading model: ``booster.save_model('model.json')``.  More details
-below.
+open format that can be easily reused.  Later in XGBoost 1.6.0, additional support for
+`Universal Binary JSON <https://ubjson.org/>`__ is added as an optimization for more
+efficient model IO.  They have the same document structure with different representations,
+and we will refere them collectively as the JSON format. This tutorial aims to share some
```
Comment (trivialfis) [inline doc/tutorials/saving_model.rst:11]: typo.

**Comment by trivialfis** [comment]
All merged as separated PRs.

**Pull request closed without merging**